server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1180,151 +916,315 @@ public void snapshot(final IndexCommit snapshotIndexCommit) {
-                // multi part file sum up the size and check
-                int part = 0;
-                long totalSize = 0;
-                while (true) {
-                    blobMetaData = blobs.get(fileInfo.partName(part++));
-                    if (blobMetaData == null) {
-                        break;
-                    }
-                    totalSize += blobMetaData.length();
-                }
-                return totalSize == fileInfo.length();
-            }
-            // no file, not exact and not multipart
-            return false;
+    @Override
+    public String toString() {
+        return "BlobStoreRepository[" +
+            "[" + metadata.name() +
+            "], [" + blobStore() + ']' +
+            ']';
+    }
+
+    /**
+     * Delete shard snapshot
+     */
+    private void deleteShardSnapshot(IndexId indexId, ShardId snapshotShardId, SnapshotId snapshotId) {
+        final BlobContainer shardContainer = shardContainer(indexId, snapshotShardId);
+        final Map<String, BlobMetaData> blobs;
+        try {
+            blobs = shardContainer.listBlobs();
+        } catch (IOException e) {
+            throw new IndexShardSnapshotException(snapshotShardId, "Failed to list content of gateway", e);
         }
 
-        private class AbortableInputStream extends FilterInputStream {
-            private final String fileName;
+        Tuple<BlobStoreIndexShardSnapshots, Integer> tuple = buildBlobStoreIndexShardSnapshots(blobs, shardContainer);
+        BlobStoreIndexShardSnapshots snapshots = tuple.v1();
+        int fileListGeneration = tuple.v2();
+
+        try {
+            indexShardSnapshotFormat.delete(shardContainer, snapshotId.getUUID());
+        } catch (IOException e) {
+            logger.warn(new ParameterizedMessage("[{}] [{}] failed to delete shard snapshot file", snapshotShardId, snapshotId), e);
+        }
 
-            AbortableInputStream(InputStream delegate, String fileName) {
-                super(delegate);
-                this.fileName = fileName;
+        // Build a list of snapshots that should be preserved
+        List<SnapshotFiles> newSnapshotsList = new ArrayList<>();
+        for (SnapshotFiles point : snapshots) {
+            if (!point.snapshot().equals(snapshotId.getName())) {
+                newSnapshotsList.add(point);
             }
+        }
+        // finalize the snapshot and rewrite the snapshot index with the next sequential snapshot index
+        finalizeShard(newSnapshotsList, fileListGeneration, blobs, "snapshot deletion [" + snapshotId + "]", shardContainer,
+            snapshotShardId, snapshotId);
+    }
 
-            @Override
-            public int read() throws IOException {
-                checkAborted();
-                return in.read();
+    /**
+     * Loads information about shard snapshot
+     */
+    private BlobStoreIndexShardSnapshot loadShardSnapshot(IndexId indexId, ShardId snapshotShardId, SnapshotId snapshotId) {
```

Comment:
We could maybe pass directly the BlobContainer to this method?

Reply:
Yea that looks much nicer :)